src/coreclr/jit/promotion.cpp

```diff
@@ -743,31 +783,551 @@ class ReplaceVisitor : public GenTreeVisitor<ReplaceVisitor>
+
+                if ((rep == firstRep) && m_compiler->fgIsBigOffset(srcOffs) &&
+                    m_compiler->fgAddrCouldBeNull(src->gtGetOp1()))
+                {
+                    GenTree* addrForNullCheck = m_compiler->gtCloneExpr(src->gtGetOp1());
+                    result->AddStatement(m_compiler->gtNewIndir(TYP_BYTE, addrForNullCheck));
+                    UpdateEarlyRefCount(addrForNullCheck);
+                }
+
+                GenTree* addr = m_compiler->gtCloneExpr(src->gtGetOp1());
+                UpdateEarlyRefCount(addr);
+                if (srcOffs != 0)
+                {
+                    var_types addrType = varTypeIsGC(addr) ? TYP_BYREF : TYP_I_IMPL;
+                    addr               = m_compiler->gtNewOperNode(GT_ADD, addrType, addr,
+                                                     m_compiler->gtNewIconNode(srcOffs, TYP_I_IMPL));
+                }
+
+                GenTree* dstLcl = m_compiler->gtNewLclvNode(rep->LclNum, rep->AccessType);
+                srcFld          = m_compiler->gtNewIndir(rep->AccessType, addr, src->gtFlags & GTF_IND_VOLATILE);
+                srcFld->gtFlags |= GTF_GLOB_REF;
+            }
+
+            result->AddStatement(m_compiler->gtNewAssignNode(dstLcl, srcFld));
+            rep->Handled = true;
+        }
+    }
+
+    //------------------------------------------------------------------------
+    // UpdateEarlyRefCount:
+    //   Update early ref counts if necessary for the specified IR node.
+    //
+    // Parameters:
+    //   candidate - the IR node that may be a local that should have its early ref counts updated.
+    //
+    void UpdateEarlyRefCount(GenTree* candidate)
+    {
+        if (!candidate->OperIs(GT_LCL_VAR, GT_LCL_FLD, GT_LCL_ADDR))
+        {
+            return;
+        }
+
+        IncrementRefCount(candidate->AsLclVarCommon()->GetLclNum());
+
+        LclVarDsc* varDsc = m_compiler->lvaGetDesc(candidate->AsLclVarCommon());
+        if (varDsc->lvIsStructField)
+        {
+            IncrementRefCount(varDsc->lvParentLcl);
         }
 
-        if (asg->gtGetOp1()->OperIs(GT_LCL_VAR, GT_LCL_FLD))
+        if (varDsc->lvPromoted)
         {
-            GenTreeLclVarCommon* lhsLcl = asg->gtGetOp1()->AsLclVarCommon();
-            if (lhsLcl->TypeIs(TYP_STRUCT))
+            for (unsigned fldLclNum = varDsc->lvFieldLclStart; fldLclNum < varDsc->lvFieldLclStart + varDsc->lvFieldCnt;
+                 fldLclNum++)
             {
-                unsigned size = lhsLcl->GetLayout(m_compiler)->GetSize();
-                MarkForReadBack(lhsLcl->GetLclNum(), lhsLcl->GetLclOffs(), size, true);
+                IncrementRefCount(fldLclNum);
             }
         }
     }
 
+    //------------------------------------------------------------------------
+    // IncrementRefCount:
+    //   Increment the ref count for the specified local.
+    //
+    // Parameters:
+    //   lclNum - the local
+    //
+    void IncrementRefCount(unsigned lclNum)
```

Comment:
Share this with local address visitor?

Also the local address visitor's `UpdateEarlyRefCount` does things with weighted ref counts which would be difficult to duplicate here -- is there any adverse interaction caused by not updating them?

Reply:
> Share this with local address visitor?

Will do.

> Also the local address visitor's UpdateEarlyRefCount does things with weighted ref counts which would be difficult to duplicate here -- is there any adverse interaction caused by not updating them?

The ref counts we are updating here are only for "existing" locals that appear under assignments, so I don't think the weighted ref counts are problematic.

The remainder of the pass is not handling the ref counts properly for the locals it introduces, I will submit a separate change to get that sorted out. There won't be an adverse effect of not updating the weighted ref count since we only use that to undo "old" promotion (the other use of them, for omitting some implicit byref copies, was replaced by liveness if you recall).

Reply:
I introduced an `LclVarDsc::incLvRefCntSaturating` and used it from both. Didn't seem too easy to factor the rest of the promotion-related common logic due to the weighted ref-count logic you mentioned.